source/MRMesh/MRColorMapAggregator.cpp

```diff
@@ -0,0 +1,171 @@
+#include "MRColorMapAggregator.h"
+#include "MRGTest.h"
+
+namespace MR
+{
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setDefaultColor( const Color& color )
+{
+    defaultColor_ = color;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::pushBack( const PartitialColorMap& partitialColorMap )
+{
+    assert( checkInputData_( partitialColorMap ) );
+    dataSet_.push_back( partitialColorMap );
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::insert( int i, const PartitialColorMap& partitialColorMap )
+{
+    assert( i <= dataSet_.size() );
+    assert( checkInputData_( partitialColorMap ) );
+    dataSet_.insert( dataSet_.begin() + i, partitialColorMap );
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::replace( int i, const PartitialColorMap& partitialColorMap )
+{
+    assert( i >= 0 && i < dataSet_.size() );
+    assert( checkInputData_( partitialColorMap ) );
+    dataSet_[i] = partitialColorMap;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::reset()
+{
+    dataSet_.clear();
+    needUpdate_ = true;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::erase( int i, int n /*= 1*/ )
+{
+    assert( i >= 0 && i + n <= dataSet_.size() );
+    dataSet_.erase( dataSet_.begin() + i, dataSet_.begin() + i + n );
+
+    needUpdate_ = true;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setMode( AggregateMode mode )
+{
+    if ( mode == mode_ )
+        return;
+    mode_ = mode;
+    needUpdate_ = true;
+}
+
+template<typename Tag>
+typename ColorMapAggregator<Tag>::ColorMap ColorMapAggregator<Tag>::aggregate( const ElementBitSet& elementBitSet )
+{
+    int last = elementBitSet.find_last();
+    if ( needUpdate_ )
+        updateAggregated_( last + 1 );
+    else if ( last >= aggregatedColorMap_.size() )
+        aggregatedColorMap_.resize( last + 1, defaultColor_ );
+
+    ColorMap res( elementBitSet.size() );
+    for ( const auto& e : elementBitSet )
+        res[e] = aggregatedColorMap_[e];
+
+    return res;
+}
+
+template<typename Tag>
+bool ColorMapAggregator<Tag>::checkInputData_( const PartitialColorMap& partitialColorMap )
+{
+    return !partitialColorMap.colorMap.empty() &&
+        ( partitialColorMap.colorMap.size() > partitialColorMap.elements.find_last() );
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::updateAggregated_( int newSize )
+{
+    aggregatedColorMap_.clear();
+    int maxSize = newSize;
+    for ( int i = 0; i < dataSet_.size(); ++i )
+        maxSize = std::max( maxSize, int( dataSet_[i].elements.find_last() ) + 1 );
+    aggregatedColorMap_.resize( maxSize, defaultColor_ );
+
+    ElementBitSet remaining;
+    remaining.resize( maxSize, true );
```

Comment:
mb move this under `AggregateMode::Overlay` block

Reply:
fixed